### PR TITLE
Serialization improvements step 3 (compression.h)

### PR DESCRIPTION
**Pull request by sipa**
This is the next piece of the puzzle from #10785. It includes:
* The `FORMATTER_METHODS` macro, similar to `SERIALIZE_METHODS`, for defining a formatter with a unified serialization/deserialization implementation.
* Updating `compression.h` to consist of 3 formatters, rather than old-style wrappers (`ScriptCompression`, `AmountCompression`, `TxOutCompression`).

### CODE REVIEW

**Comment by sipa** [comment]
Ping @ryanofsky, as you volunteered to review further PRs in this series. I've made an additional change compared to the old big PR, namely the introduction of FORMATTER_METHODS (and then rewriting SERIALIZE_METHODS in function of FORMATTER_METHODS, which follows pretty naturally).

**Comment by practicalswift** [comment]
https://github.com/bitcoin/bitcoin/blob/fd691da3b032e439936672e084b7c49086ec9ebc/src/coins.h#L59-L65

A signed integer overflow will take place on L62 in case of serialising a coin with a very large `nHeight` value.

It wasn't introduced in this PR, but since you're making changes to this function it might be worth fixing it :)

Also the `Unserialize` function should probably be adjusted to not accept `block_height`s above `(std::numeric_limits<uint32_t>::max() - 1) / 2`.

**Comment by sipa** [comment]
@practicalswift How about we address that when coins.h itself is updated to the new serialization framework (here it's just adjusting for changes in compression.h)?

File: src/undo.h [hunk 1]
```diff
@@ -32,7 +32,7 @@ class TxInUndoSerializer
             // Required to maintain compatibility with older undo format.
             ::Serialize(s, (unsigned char)0);
         }
-        ::Serialize(s, CTxOutCompressor(REF(txout->out)));
+        ::Serialize(s, Using<TxOutCompression>(REF(txout->out)));
```
Comment (ryanofsky) [inline src/undo.h]: In commit "Convert compression.h to new serialization framework" (fd691da3b032e439936672e084b7c49086ec9ebc)

Can you drop uses of `REF` macro here and below?

File: src/serialize.h [hunk 2]
```diff
@@ -211,17 +235,15 @@ template<typename X> const X& ReadWriteAsHelper(const X& x) { return x; }
...
     void Unserialize(Stream& s)                                                     \
     {                                                                               \
         static_assert(std::is_same<cls&, decltype(*this)>::value, "Unserialize type mismatch"); \
-        SerializationOps(*this, s, CSerActionUnserialize());                        \
+        Unser(s, *this);                                                            \
     }                                                                               \
-    template<typename Stream, typename Type, typename Operation>                    \
-    static inline void SerializationOps(Type& obj, Stream& s, Operation ser_action) \
-
+    FORMATTER_METHODS(cls, obj)
```
Comment (ryanofsky) [inline src/serialize.h]: In commit "Add FORMATTER_METHODS, similar to SERIALIZE_METHODS, but for formatters" (ca34c5cba5fbb9b046b074a234f06ecf6ed5d610)

It's a little surprising to see SERIALIZE_METHODS depend on FORMATTER_METHODS, but I guess it saves some repetition, and it seems reasonable

**Review by ryanofsky — APPROVED** [review]
Code review ACK fd691da3b032e439936672e084b7c49086ec9ebc

File: src/undo.h (hunk 1 above)
Comment (sipa) [reply src/undo.h]: Indeed, done.

File: src/serialize.h (hunk 2 above)
Comment (sipa) [reply src/serialize.h]: I think it makes sense: every class gets the ability to function as a formatter. Both FORMATTER_METHODS and SERIALIZE_METHODS help do that; with the first it's for an arbitrary type, with the latter it's for objects of the class it's being defined it. The latter then also hooks up the Serialize/Unserialize functions to the self-formatter.

Do you suggest documenting this better, or have some other suggestion?

Comment (ryanofsky) [reply src/serialize.h]: > > It's a little surprising to see SERIALIZE_METHODS depend on FORMATTER_METHODS, but I guess it saves some repetition, and it seems reasonable

> I think it makes sense: every class gets the ability to function as a formatter. Both FORMATTER_METHODS and SERIALIZE_METHODS help do that; with the first it's for an arbitrary type, with the latter it's for objects of the class it's being defined it. The latter then also hooks up the Serialize/Unserialize functions to the self-formatter.

I think it's fine. A formatter class with a formatter methods is a class that can serialize and deserialize objects in a particular format. A serializable type with serialize methods is a class whose instances can serialize and deserialize themselves. I wasn't expecting serializeable objects to now be able to format other objects of the same type, and I couldn't think of a use for that, but I think it's harmless.

> Do you suggest documenting this better, or have some other suggestion?

I guess the suggestion would be to keep the new FORMATTER_METHODS macro exactly like you have it here, but revert the changes to the SERIALIZE_METHODS macro, so it isn't affected by this PR. But only if you want to have a stronger distinction between formatter classes and serializable types. I'm fine either way.

**Review by ryanofsky — APPROVED** [review]
Code review ACK 4de934b9b5b4be1bac8fe205f4ee9a79e772dc34. Only change since last review is removing REF usages

**Comment by laanwj** [comment]
code review ACK 4de934b9b5b4be1bac8fe205f4ee9a79e772dc34

**Pull request merged**